File: go/controlled_generation.go
package examples

import (
	"context"
	"path/filepath"
	"os"
	"log"

	"google.golang.org/genai"
)

func JsonControlledGeneration() (*genai.GenerateContentResponse, error) {
	// [START json_controlled_generation]
	ctx := context.Background()
	client, err := genai.NewClient(ctx, &genai.ClientConfig{
		APIKey:  os.Getenv("GEMINI_API_KEY"), 
		Backend: genai.BackendGeminiAPI,
	})
	if err != nil {
		log.Fatal(err)
	}

	schema := &genai.Schema{
		Type: genai.TypeArray,
		Items: &genai.Schema{
			Type: genai.TypeObject,
			Properties: map[string]*genai.Schema{
				"recipe_name": {Type: genai.TypeString},
				"ingredients": {
					Type:  genai.TypeArray,
					Items: &genai.Schema{Type: genai.TypeString},
				},
			},
			Required: []string{"recipe_name"},
		},
	}

	config := &genai.GenerateContentConfig{
		ResponseMIMEType: "application/json",
		ResponseSchema:   schema,
	}

	response, err := client.Models.GenerateContent(
		ctx,
		"gemini-3.8-flash",
		genai.Text("List a few popular cookie recipes."),
		config,
	)
	if err != nil {
		log.Fatal(err)
	}
	printResponse(response)
	// [END json_controlled_generation]
	return response, err
}

func JsonNoSchema() (*genai.GenerateContentResponse, error) {
	// [START json_no_schema]
	ctx := context.Background()
	client, err := genai.NewClient(ctx, &genai.ClientConfig{
		APIKey:  os.Getenv("GEMINI_API_KEY"), 
		Backend: genai.BackendGeminiAPI,
	})
	if err != nil {
		log.Fatal(err)
	}
	prompt := "List a few popular cookie recipes in JSON format.\n\n" +
			  "Use this JSON schema:\n\n" +
			  "Recipe = {'recipe_name': str, 'ingredients': list[str]}\n" +
		      "Return: list[Recipe]"
	response, err := client.Models.GenerateContent(ctx, "gemini-3.8-flash", genai.Text(prompt), nil)
	if err != nil {
		log.Fatal(err)
	}
	printResponse(response)
	// [END json_no_schema]
	return response, err
}

func JsonEnum() (*genai.GenerateContentResponse, error) {
	// [START json_enum]
	ctx := context.Background()
	client, err := genai.NewClient(ctx, &genai.ClientConfig{
		APIKey:  os.Getenv("GEMINI_API_KEY"), 
		Backend: genai.BackendGeminiAPI,
	})
	if err != nil {
		log.Fatal(err)
	}

	// Choice is a custom type representing a musical instrument category.
	type Choice string

	const (
		Percussion Choice = "Percussion"
		String     Choice = "String"
		Woodwind   Choice = "Woodwind"
		Brass      Choice = "Brass"
		Keyboard   Choice = "Keyboard"
	)

	// Define a schema restricting the response to the allowed Choice enum values.
	schema := &genai.Schema{
		Type: genai.TypeString,
		Enum: []string{
			string(Percussion),
			string(String),
			string(Woodwind),
			string(Brass),
			string(Keyboard),
		},
	}

	config := &genai.GenerateContentConfig{
		ResponseMIMEType: "application/json",
		ResponseSchema:   schema,
	}

	file, err := client.Files.UploadFromPath(
		ctx, 
		filepath.Join(getMedia(), "organ.jpg"), 
		&genai.UploadFileConfig{
			MIMEType : "image/jpeg",
		},
	)
	if err != nil {
		log.Fatal(err)
	}
	parts := []*genai.Part{
		genai.NewPartFromText("What kind of instrument is this:"),
		genai.NewPartFromURI(file.URI, file.MIMEType),
	}
	contents := []*genai.Content{
		genai.NewContentFromParts(parts, genai.RoleUser),
	}

	response, err := client.Models.GenerateContent(ctx, "gemini-3.8-flash",
		contents,
		config,
	)
	if err != nil {
		log.Fatal(err)
	}
	printResponse(response)
	// [END json_enum]
	return response, err
}

func EnumInJson() (*genai.GenerateContentResponse, error) {
	// [START enum_in_json]
	ctx := context.Background()
	client, err := genai.NewClient(ctx, &genai.ClientConfig{
		APIKey:  os.Getenv("GEMINI_API_KEY"), 
		Backend: genai.BackendGeminiAPI,
	})
	if err != nil {
		log.Fatal(err)
	}

	// We use a schema representing an array of objects.
	schema := &genai.Schema{
		Type: genai.TypeArray,
		Items: &genai.Schema{
			Type: genai.TypeObject,
			Properties: map[string]*genai.Schema{
				"recipe_name": {Type: genai.TypeString},
				"grade":       {Type: genai.TypeString, Enum: []string{"a+", "a", "b", "c", "d", "f"}},
			},
			Required: []string{"recipe_name", "grade"},
		},
	}
	config := &genai.GenerateContentConfig{
		ResponseMIMEType: "application/json",
		ResponseSchema:   schema,
	}
	response, err := client.Models.GenerateContent(ctx, "gemini-3.8-flash",
		genai.Text("List about 10 cookie recipes, grade them based on popularity"),
		config,
	)
	if err != nil {
		log.Fatal(err)
	}
	// Expected output: a JSON-parsed list with recipe names and grades (e.g., "a+")
	printResponse(response)
	// [END enum_in_json]
	return response, err
}

func JsonEnumRaw() (*genai.GenerateContentResponse, error) {
	// [START json_enum_raw]
	ctx := context.Background()
	client, err := genai.NewClient(ctx, &genai.ClientConfig{
		APIKey:  os.Getenv("GEMINI_API_KEY"), 
		Backend: genai.BackendGeminiAPI,
	})
	if err != nil {
		log.Fatal(err)
	}

	config := &genai.GenerateContentConfig{
		ResponseMIMEType: "application/json",
		ResponseSchema: &genai.Schema{
			Type: genai.TypeString,
			Enum: []string{"Percussion", "String", "Woodwind", "Brass", "Keyboard"},
		},
	}

	file, err := client.Files.UploadFromPath(
		ctx, 
		filepath.Join(getMedia(), "organ.jpg"), 
		&genai.UploadFileConfig{
			MIMEType : "image/jpeg",
		},
	)
	if err != nil {
		log.Fatal(err)
	}
	parts := []*genai.Part{
		genai.NewPartFromText("What kind of instrument is this:"),
		genai.NewPartFromURI(file.URI, file.MIMEType),
	}
	contents := []*genai.Content{
		genai.NewContentFromParts(parts, genai.RoleUser),
	}

	response, err := client.Models.GenerateContent(ctx, "gemini-3.8-flash",
		contents,
		config,
	)
	if err != nil {
		log.Fatal(err)
	}

	printResponse(response)
	// [END json_enum_raw]
	return response, err
}

func XEnum() (*genai.GenerateContentResponse, error) {
	// [START x_enum]
	ctx := context.Background()
	client, err := genai.NewClient(ctx, &genai.ClientConfig{
		APIKey:  os.Getenv("GEMINI_API_KEY"),
		Backend: genai.BackendGeminiAPI,
	})
	if err != nil {
		log.Fatal(err)
	}

	// Choice is a custom type representing a musical instrument category.
	type Choice string

	const (
		Percussion Choice = "Percussion"
		String     Choice = "String"
		Woodwind   Choice = "Woodwind"
		Brass      Choice = "Brass"
		Keyboard   Choice = "Keyboard"
	)

	// Define a schema restricting the response to the allowed Choice enum values.
	schema := &genai.Schema{
		Type: genai.TypeString,
		Enum: []string{
			string(Percussion),
			string(String),
			string(Woodwind),
			string(Brass),
			string(Keyboard),
		},
	}

	config := &genai.GenerateContentConfig{
		ResponseMIMEType: "text/x.enum",
		ResponseSchema:   schema,
	}

	file, err := client.Files.UploadFromPath(
		ctx, 
		filepath.Join(getMedia(), "organ.jpg"), 
		&genai.UploadFileConfig{
			MIMEType : "image/jpeg",
		},
	)
	if err != nil {
		log.Fatal(err)
	}
	parts := []*genai.Part{
		genai.NewPartFromText("What kind of instrument is this:"),
		genai.NewPartFromURI(file.URI, file.MIMEType),
	}
	contents := []*genai.Content{
		genai.NewContentFromParts(parts, genai.RoleUser),
	}
	response, err := client.Models.GenerateContent(ctx, "gemini-3.8-flash",
		contents,
		config,
	)
	if err != nil {
		log.Fatal(err)
	}
	printResponse(response)
	// [END x_enum]
	return response, err
}

func XEnumRaw() (*genai.GenerateContentResponse, error) {
	// [START x_enum_raw]
	ctx := context.Background()
	client, err := genai.NewClient(ctx, &genai.ClientConfig{
		APIKey:  os.Getenv("GEMINI_API_KEY"),
		Backend: genai.BackendGeminiAPI,
	})
	if err != nil {
		log.Fatal(err)
	}

	rawSchema := &genai.Schema{
		Type: genai.TypeString,
		Enum: []string{"Percussion", "String", "Woodwind", "Brass", "Keyboard"},
	}
	config := &genai.GenerateContentConfig{
		ResponseMIMEType: "text/x.enum",
		ResponseSchema:   rawSchema,
	}

	file, err := client.Files.UploadFromPath(
		ctx, 
		filepath.Join(getMedia(), "organ.jpg"), 
		&genai.UploadFileConfig{
			MIMEType : "image/jpeg",
		},
	)
	if err != nil {
		log.Fatal(err)
	}
	parts := []*genai.Part{
		genai.NewPartFromText("What kind of instrument is this:"),
		genai.NewPartFromURI(file.URI, file.MIMEType),
	}
	contents := []*genai.Content{
		genai.NewContentFromParts(parts, genai.RoleUser),
	}
	response, err := client.Models.GenerateContent(ctx, "gemini-3.8-flash",
		contents,
		config,
	)
	if err != nil {
		log.Fatal(err)
	}
	printResponse(response)
	// Expected output: "Keyboard"
	// [END x_enum_raw]
	return response, err
}

File: java/src/main/java/com/example/gemini/ControlledGeneration.java
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.gemini;

import com.google.genai.Client;
import com.google.genai.types.Content;
import com.google.genai.types.GenerateContentConfig;
import com.google.genai.types.GenerateContentResponse;
import com.google.genai.types.Part;
import com.google.genai.types.Schema;
import org.jspecify.annotations.Nullable;

import java.lang.reflect.Array;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

import static com.example.gemini.BuildConfig.media_path;

public class ControlledGeneration {
    public static @Nullable String jsonControlledGeneration() {
        // [START json_controlled_generation]
        Client client = new Client();

        Schema recipeSchema = Schema.builder()
                .type(Array.class.getSimpleName())
                .items(Schema.builder()
                        .type(Object.class.getSimpleName())
                        .properties(
                                Map.of("recipe_name", Schema.builder()
                                                .type(String.class.getSimpleName())
                                                .build(),
                                        "ingredients", Schema.builder()
                                                .type(Array.class.getSimpleName())
                                                .items(Schema.builder()
                                                        .type(String.class.getSimpleName())
                                                        .build())
                                                .build())
                        )
                        .required(List.of("recipe_name", "ingredients"))
                        .build())
                .build();

        GenerateContentConfig config =
                GenerateContentConfig.builder()
                        .responseMimeType("application/json")
                        .responseSchema(recipeSchema)
                        .build();

        GenerateContentResponse response =
                client.models.generateContent(
                        "gemini-3.8-flash",
                        "List a few popular cookie recipes.",
                        config);

        System.out.println(response.text());
        // [END json_controlled_generation]
        return response.text();
    }

    public static @Nullable String jsonNoSchema() {
        // [START json_no_schema]
        Client client = new Client();

        String prompt = """
                List a few popular cookie recipes in JSON format.
                
                
                Use this JSON schema:
                
                
                Recipe = {'recipe_name': string, 'ingredients': list[string]}
                
                Return: list[Recipe].
                """;

        GenerateContentResponse response =
                client.models.generateContent(
                        "gemini-3.8-flash",
                        prompt,
                        null);

        System.out.println(response.text());
        // [END json_no_schema]
        return response.text();
    }

    public static @Nullable String jsonEnum() throws Exception {
        // [START json_enum]
        Client client = new Client();

        Schema schema =
                Schema.builder()
                        .type(String.class.getSimpleName())
                        .enum_(Arrays.asList("Percussion", "String", "Woodwind", "Brass", "Keyboard"))
                        .build();

        String path = media_path + "organ.jpg";
        byte[] imageData = Files.readAllBytes(Paths.get(path));

        Content content =
                Content.fromParts(
                        Part.fromText("What kind of instrument is this:"),
                        Part.fromBytes(imageData, "image/jpeg"));

        GenerateContentConfig config =
                GenerateContentConfig.builder()
                        .responseMimeType("application/json")
                        .candidateCount(1)
                        .responseSchema(schema)
                        .build();

        GenerateContentResponse response =
                client.models.generateContent(
                        "gemini-3.8-flash",
                        content,
                        config);

        System.out.println(response.text());
        // [END json_enum]
        return response.text();
    }

    public static @Nullable String enumInJson() throws Exception {
        // [START enum_in_json]
        Client client = new Client();

        Schema recipeSchema = Schema.builder()
                .type(Array.class.getSimpleName())
                .items(Schema.builder()
                        .type(Object.class.getSimpleName())
                        .properties(
                                Map.of("recipe_name", Schema.builder()
                                                .type(String.class.getSimpleName())
                                                .build(),
                                        "grade", Schema.builder()
                                                .type(String.class.getSimpleName())
                                                .enum_(Arrays.asList("a+", "a", "b", "c", "d", "f"))
                                                .build())
                        )
                        .required(List.of("recipe_name", "grade"))
                        .build())
                .build();

        GenerateContentConfig config =
                GenerateContentConfig.builder()
                        .responseMimeType("application/json")
                        .responseSchema(recipeSchema)
                        .build();

        GenerateContentResponse response =
                client.models.generateContent(
                        "gemini-3.8-flash",
                        "List about 10 cookie recipes, grade them based on popularity",
                        config);

        System.out.println(response.text());
        // [END enum_in_json]
        return response.text();
    }

    public static @Nullable String xEnum() throws Exception {
        // [START x_enum]
        Client client = new Client();

        Schema schema =
                Schema.builder()
                        .type(String.class.getSimpleName())
                        .enum_(Arrays.asList("Percussion", "String", "Woodwind", "Brass", "Keyboard"))
                        .build();

        String path = media_path + "organ.jpg";
        byte[] imageData = Files.readAllBytes(Paths.get(path));

        Content content =
                Content.fromParts(
                        Part.fromText("What kind of instrument is this:"),
                        Part.fromBytes(imageData, "image/jpeg"));

        GenerateContentConfig config =
                GenerateContentConfig.builder()
                        .responseMimeType("text/x.enum")
                        .candidateCount(1)
                        .responseSchema(schema)
                        .build();

        GenerateContentResponse response =
                client.models.generateContent(
                        "gemini-3.8-flash",
                        content,
                        config);

        System.out.println(response.text());
        // [END x_enum]
        return response.text();
    }
}
